File: Lab/Lab5/Lab5.java
package Lab.Lab5;
import java.util.*;
public class Lab5 {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(3);
        root.left.left = new TreeNode(1);
        //root.left.right = new TreeNode(4);
        root.right = new TreeNode(8);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);
        System.out.println(isValidBST(root));
    }
    public static boolean isValidBST(TreeNode root){
        return validHelper(root.left, Integer.MIN_VALUE, root.val) && validHelper(root.right, root.val, Integer.MAX_VALUE);
    }
    public static boolean validHelper(TreeNode root, int lower, int higher){
        if(root == null ){
            return true;
        }
        if(root.val >= higher || root.val <= lower){
            return false;
        }
        return validHelper(root.left, lower, root.val) && validHelper(root.right, root.val, higher);
    }
}
class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: Lab/Lab5/Lab5.py
import sys
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
def isValidBST(root):
    return isValidHelper(root.left, -sys.maxsize - 1, root.val) and isValidHelper(root.left, root.val, sys.maxsize)
def isValidHelper(root, lower, upper):
    if(root is None):
        return True
    if not (lower < root.val < upper):
        return False
    return isValidHelper(root.left, lower, root.val) and isValidHelper(root.right, root.val, upper)

def main():
    root = TreeNode(4)
    root.left = TreeNode(3)
    root.left.left = TreeNode(1)
    root.right = TreeNode(8)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(9)
    print(isValidBST(root))

if __name__ == "__main__":
    main()
    

    
